### PR TITLE
Avoid throw away functions and eval trickery

### PR DESCRIPTION
**Pull request by jdalton**
Related to the discussion at https://github.com/sindresorhus/to-fast-properties/pull/7#issuecomment-298243859.

I cleaned up the method to avoid throw away functions and eval trickery.
I also documented the technique a bit more.

### CODE REVIEW

**Comment by benjamingr** [comment]
The eval trickery is just to guard at this point against future dead code elimination instructions. It is not actually required in the code.

The existing code can be modified to:


```js
function toFastProperties(o) {
  const s = {__proto__: o};
  const ic = () => s.foo;
  ic();
  ic();
  return o;
}
```

Which is shorter and does not require another class. It can be further shortened, but at the expense of readability.

**Comment by jdalton** [comment]
> The eval trickery is just to guard at this point against future dead code elimination instructions.

Guarding against future whatever isn't so great as the heuristic can and did shift out from underneath the package. It's superstitious coding, which I'm not a fan of. The best strategy is to simply keep the test matrix up to date.

> Which is shorter and does not require another class.

Your new suggestion still has throw away functions, adds a throw away object, and relies on the special `__proto__` which I tried to avoid. This PR could move the body of `FastObject` into `toFastProperties` but I kept it out for readability.

File: index.js [hunk 1]
```diff
@@ -1,13 +1,27 @@
...
+function FastObject(o) {
+  // A prototype object will have "fast properties" enabled once it is checked
+  // against the inline property cache of a function, e.g. fastProto.property:
+  // https://github.com/v8/v8/blob/6.0.122/test/mjsunit/fast-prototype.js#L48-L63
+  if (fastProto !== null && typeof fastProto.property) {
+    const result = fastProto;
+    fastProto = FastObject.prototype = null;
+    return result;
+  }
+  fastProto = FastObject.prototype = o == null ? Object.create(null) : o;
```
Comment (jdalton) [inline index.js]: Bonus: Creates a fast object if one is not provided.

**Comment by benjamingr** [comment]
@jdalton note I'm not guarding against a future V8 optimization here, I'm guarding against minifiers - if the code is used in a client library, it is likely the `eval` will cause the function to not be removed (as dead code) whereas if the eval isn't there then the function has no effects and can be removed.

**Comment by jdalton** [comment]
I'm glad you brought that up. I actually tested this PR, before opening, against JIT dead code removal and against closure compiler (regular and advanced) and uglifyjs and it's fine.

**Comment by sindresorhus** [comment]
So, are we ok with merging this?

**Comment by benjamingr** [comment]
I'm ok with merging this, though I don't think it offers any advantage over the current implementation - the fact @jdalton feels strongly about it is enough for me. 

Also, if he's adding himself to the credits/copyright please add me too :)

**Comment by sindresorhus** [comment]
Cool :)

**Comment by sindresorhus** [comment]
> Also, if he's adding himself to the credits/copyright please add me too :)

Definitely. Can you add yourself?

**Pull request merged**